File: 15-01-2025_Longest_Subarray_with_Sum_K.cpp
class Solution {
  public:
    int longestSubarray(vector<int>& arr, int k) 
    {
        int n = arr.size();
    
        unordered_map <int,int> mp ;
        
        int ans = 0;
        
        int psum =0;
        
        for(int i=0;i<n;i++)
        {
            psum+=arr[i];
            
            if(psum == k)
            {
                ans = i+1;
            }
            
            
            if(mp.find(psum - k)!=mp.end())
            {
                ans = max(ans,i - mp[psum-k]);
            }
            
            if(mp.find(psum)==mp.end())
            {
                mp[psum]=i;
            }
        }
        
        
        return ans;
        
     
    }
};

File: 15-01-2025_Longest_Subarray_with_Sum_K.java
class Solution {
    public int longestSubarray(int[] arr, int k) {
        int n = arr.length;

        Map<Integer, Integer> map = new HashMap<>();
        int ans = 0;
        int psum = 0;

        for (int i = 0; i < n; i++) {
            psum += arr[i];

            if (psum == k) {
                ans = i + 1;
            }

            if (map.containsKey(psum - k)) {
                ans = Math.max(ans, i - map.get(psum - k));
            }

            if (!map.containsKey(psum)) {
                map.put(psum, i);
            }
        }

        return ans;
    }
}
